File: 11.container-with-most-water.java
/*
 * @lc app=leetcode id=11 lang=java
 *
 * [11] Container With Most Water
 */

// @lc code=start
class Solution {
    public int maxArea(int[] height) {
        int n = height.length;
        int left = 0;
        int right = n - 1;
        int maxArea = 0;
        while (left < right) {
            int currentArea = Math.min(height[left], height[right]) * (right - left);
            maxArea = Math.max(maxArea, currentArea);
            if (height[left] < height[right]) {
                int preleft = height[left];
                while (height[left] <= preleft && left < right) {
                    left++;
                }
            } else {
                int preright = height[right];
                while (height[right] <= preright && left < right) {
                    right--;
                }
            }
        }

        return maxArea;
    }
}
// @lc code=end

File: 11.container-with-most-water.py
#
# @lc app=leetcode id=11 lang=python3
#
# [11] Container With Most Water
#

# @lc code=start


class Solution:
    def maxArea(self, height: List[int]) -> int:
        n=len(height)
        left, right=0,n-1
        max_area=0
        while left<right:
            curr_area=min(height[left], height[right])*(right-left)
            max_area=max(max_area, curr_area)
            # how do we walk throuh this?
            if height[left]<= height[right]:
                left+=1
            else:
                right-=1

        return max_area

# @lc code=end

